File: bindings/forth/safe-unit.fs
\ SPDX-License-Identifier: PMPL-1.0-or-later
\ Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
\
\ Proven SafeUnit - FFI bindings to libproven unit conversion operations.
\ All unit conversions are performed in verified Idris 2 code via libproven.

c-library proven_unit
s" proven" add-lib

\ ProvenFloatResult is { int32_t status; double value; }
c-function proven-length-conv proven_unit_convert_length r n n -- a
c-function proven-temp-conv   proven_unit_convert_temp   r n n -- a

end-c-library

\ Length unit constants (matching ProvenLengthUnit enum)
0 constant LENGTH-METERS
1 constant LENGTH-KILOMETERS
2 constant LENGTH-CENTIMETERS
3 constant LENGTH-MILLIMETERS
4 constant LENGTH-FEET
5 constant LENGTH-INCHES
6 constant LENGTH-MILES
7 constant LENGTH-YARDS

\ Temperature unit constants (matching ProvenTempUnit enum)
0 constant TEMP-CELSIUS
1 constant TEMP-FAHRENHEIT
2 constant TEMP-KELVIN

\ Convert length between units via libproven
\ ( F: value -- F: result ) ( from-unit to-unit -- status )
: convert-length ( F: value -- F: result ; from to -- status )
    proven-length-conv ;

\ Convert temperature between units via libproven
\ ( F: value -- F: result ) ( from-unit to-unit -- status )
: convert-temp ( F: value -- F: result ; from to -- status )
    proven-temp-conv ;

File: bindings/purescript/Proven/SafeUnit.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// SafeUnit FFI implementation - delegates to libproven via ffi-napi.
// All computation performed in Idris 2 via the Zig FFI layer.

"use strict";

const { lib } = require("./FFI.js");

// LengthUnit enum values matching Zig FFI
var LengthUnit = {
  meters: 0,
  feet: 1,
  inches: 2,
  centimeters: 3,
  millimeters: 4,
  kilometers: 5,
  miles: 6,
  yards: 7,
};

// TempUnit enum values matching Zig FFI
var TempUnit = {
  celsius: 0,
  fahrenheit: 1,
  kelvin: 2,
};

exports.convertLengthImpl = function (value) {
  return function (from) {
    return function (to) {
      var r = lib.proven_unit_convert_length(value, from, to);
      return { status: r.status, value: r.value };
    };
  };
};

exports.convertTempImpl = function (value) {
  return function (from) {
    return function (to) {
      var r = lib.proven_unit_convert_temp(value, from, to);
      return { status: r.status, value: r.value };
    };
  };
};

exports.lengthUnits = LengthUnit;
exports.tempUnits = TempUnit;
